### PR TITLE
fix: prevent crash when logging malformed error objects

### LINKED ISSUE #60: Library Crashes When Logging Malformed Error Objects
# Library Crashes When Logging Malformed Error Objects

## Summary

The guardrails library crashes with `TypeError: Cannot read properties of undefined (reading 'value')` when `console.error()` attempts to log error objects. Node's `util.inspect` fails when serializing malformed error structures.

## Environment

- **Package**: `@openai/guardrails` version `0.2.1`
- **Node.js**: `24.12.0`
- **OS**: macOS (darwin 24.4.0)

## Error

```
Error running guardrail Jailbreak: TypeError: Cannot read properties of undefined (reading 'value')
    at formatProperty (node:internal/util/inspect:2280:12)
    at formatRaw (node:internal/util/inspect:1176:9)
    at formatValue (node:internal/util/inspect:932:10)
    at inspect (node:internal/util/inspect:409:10)
    at formatWithOptionsInternal (node:internal/util/inspect:2645:40)
    at formatWithOptions (node:internal/util/inspect:2507:10)
    at console.value (node:internal/console/constructor:377:14)
    at console.error (node:internal/console/constructor:444:61)
    at runLLM (src/checks/llm-base.ts:453:13)
```

## Problem

In `src/checks/llm-base.ts`, the `runLLM` function calls console methods with error objects:

```typescript
console.error('LLM guardrail failed for prompt:', systemPrompt, error);
console.warn('Content filter triggered by provider:', error);
console.warn('LLM returned non-JSON or malformed JSON.', error);
console.warn('LLM response validation failed.', error);
```

When `error` has malformed structures (e.g., empty objects or circular references), `util.inspect` crashes while formatting the error for logging.

**Note**: All `console.error()` and `console.warn()` calls in the catch block are potentially affected, not just the one at line 453.

## Reproduction

1. Configure a Jailbreak guardrail with a model that causes validation failures (e.g., `gpt-4.1-mini`)
2. Make an API call using the guardrails client
3. When the guardrail's internal LLM call fails validation, the library attempts to log the error
4. The crash occurs when `console.error()` or `console.warn()` tries to format the malformed error object

## Potenial Fix?

Use safe error serialization or wrap console calls in try-catch:

```typescript
const safeErrorString = error instanceof Error ? error.message : JSON.stringify(error, null, 2);
console.error('LLM guardrail failed:', safeErrorString);
```

### PR DESCRIPTION
**Pull request by LuciferDono**
## Summary

- Add `safeErrorString()` helper that safely converts errors to loggable strings without crashing Node's `util.inspect`
- Replace all raw `error` references in `console.error`/`console.warn` calls and `String(error)` conversions in `src/checks/llm-base.ts` with the safe helper
- Add test verifying null-prototype error objects no longer crash the library

## Problem

When `runLLM` catches an error with a malformed structure (e.g. empty objects, null-prototype objects, or circular references), passing it directly to `console.error()` crashes Node's `util.inspect` with `TypeError: Cannot read properties of undefined (reading 'value')`. Similarly, `String(error)` crashes with `Cannot convert object to primitive value` on null-prototype objects.

This affects all error logging paths in the catch block (lines 453, 457, 474, 490) and error message serialization (lines 479, 525).

## Fix

A small `safeErrorString()` function:
1. `Error` instances → use `error.stack` or `error.message`
2. Other objects → `JSON.stringify()` with try-catch fallback to `String()`

## Test plan

- [x] Added unit test throwing `Object.create(null)` — verifies no crash, fail-open preserved
- [x] All 349 existing tests pass
- [x] TypeScript compiles cleanly

Fixes #60

### CODE REVIEW

**Review by HAYDEN-OAI — APPROVED** [review]
Reviewed `5bd994b22cd970e254c93dc90fda1c8a741932bf`.

Reviewed the error serialization helper and its LLM failure-handling call sites. No actionable introduced correctness or compatibility issues found.

Source review only; tests were not executed, and CI status and mergeability were outside the review scope.

**Comment by jbeckwith-oai** [comment]
Thanks for tracking this down and putting together the fix! Centralizing error formatting is a useful improvement, and the null-prototype regression test covers the reported string-conversion problem.

I reviewed the current commit (`5bd994b`) in detail. Could you address these points before we merge?

1. **Make formatting safe all the way through.** `safeErrorString()` can still throw: `instanceof` and stack/message property access are outside the `try`, and the fallback `String(error)` is unprotected. Guard these operations and use a constant string as the final fallback. Read the stack once into a local variable before checking and returning it, so an accessor cannot return a different value after validation.
2. **Preserve returned error messages.** Using the logging formatter for `info.error_message` adds JSON quotes to provider strings and replaces ordinary `Error` messages with full stacks. Preserve the existing `String(error)` format where conversion succeeds, with protected fallbacks for malformed values. Stack details can remain in logs. Ensure the helper always returns a string, including for `undefined` and symbols.
3. **Protect error classification too.** The catch block's `instanceof` and `constructor.name` checks can themselves fail on malformed values. If classification fails, return the existing generic failure result rather than raising a second exception.
4. **Extend the focused tests and add a patch changeset.** Cover failed string/JSON conversion, inaccessible or changing stack accessors, inaccessible constructor properties, ordinary error-message compatibility, and provider content-filter messages/flags. Add a `.changeset/*.md` entry for `@openai/guardrails` describing the user-visible fix. The current test comment should describe null-prototype string conversion rather than claim that this object itself crashes `util.inspect`.

I tested a local correction limited to these points: all 358 tests, the TypeScript build, and lint passed, and two consecutive independent review rounds found no remaining blockers. I couldn't push that correction to the contributor fork from this environment, so those results apply to my local changes, not the current PR commit.

Please update this PR when you have a chance, and we can recheck the final diff and CI. Thanks again for the contribution!

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
comments inline

File: src/checks/llm-base.ts [hunk 1]
```diff
@@ -507,7 +522,7 @@ export async function runLLM<TOutput extends ZodTypeAny>(
         flagged: false,
         confidence: 0.0,
         info: {
-          error_message: String(error),
+          error_message: safeErrorString(error),
```
Comment (HAYDEN-OAI) [inline src/checks/llm-base.ts:525]: [P2] Preserve the public error-message representation

Using the logging formatter here changes caller-visible `info.error_message`: ordinary `Error` values now include the full stack rather than the previous `String(error)` result, and the provider-filter branch similarly adds JSON quotes to string errors. `createErrorResult` forwards this value into returned diagnostics and `originalException`, so this changes the failure contract for otherwise ordinary errors. As noted in the existing review feedback, preserve successful `String(error)` conversion for returned messages, with protected fallbacks for malformed values, and keep log formatting separate.

**Review by HAYDEN-OAI — CHANGES_REQUESTED** [review]
Re-reviewed the current head and confirmed the outstanding error-handling feedback. In particular, the returned error-message regression below means my earlier approval should not stand. The remaining feedback in https://github.com/openai/openai-guardrails-js/pull/64#issuecomment-5619799610 also needs addressing. Source review only; tests were not run.
